### ISSUE TITLE
Investigate memory leak

### ISSUE THREAD

**Issue by jpdillingham** [slskd/slskd#553]
![image](https://user-images.githubusercontent.com/17145758/179869816-22ddb05b-6744-496f-b42e-6488ad6f0b46.png)

I'm going to have to recreate this, then remote into the container and get a memory dump to see what it is.

**Comment by jpdillingham** [comment]
#557 adds a way to retrieve memory dumps easily; once the canary with that change is out I'll update and see if I can catch it in the act.  It should be easy to resolve once I figure out what it is.

**Comment by jpdillingham** [comment]
I disabled the distributed network, which accounted for a large drop in memory usage.  It then increased steadily and began rising and falling:

![image](https://user-images.githubusercontent.com/17145758/181612203-38a4f439-9033-44bb-a65c-445c75c98958.png)

It's been like this for a few days and I've uploaded a couple dozen files over that time.  With this result, I don't think the memory leak is in servicing external search requests or processing uploads.  I'll proceed by running a few searches and downloading a few files.

I suspect that root cause is in the distributed network, as it has been in the past, but I have changed a few things and need to do due dilligence.

**Comment by ckbaudio** [comment]
Having the same issues here 

![image](https://user-images.githubusercontent.com/73704121/181630618-46c9c97e-d11b-4b0a-bb40-96e358bab07e.png)

**Comment by jpdillingham** [comment]
@ckbaudio thanks for letting me know!  Could you tell me how many children you currently have? 

Linking #251; this was the last time I went down this rabbit hole.

Stats for the last 6 hours:

![image](https://user-images.githubusercontent.com/17145758/181641112-49ebd8ea-f1e1-481f-abe9-f8d7047900de.png)

The hump in the middle was about 5 searches and downloads of a couple dozen files.  I'll let this run as is until morning, but as of right now it looks like the trend returned to the mean after that activity.

I'm realizing that with the distributed network completely disabled, I'm not servicing any search requests, so that could still be the culprit.  I'll re-enable it but disallow children.

**Comment by ckbaudio** [comment]
I'd be interested to see how it performs without child processes while searches are enabled. At it's peak, it had 30+ childs, oldest spawn being 8 hours old. I'm not sure that it's searches, but it may be due to upload handles?- I don't know if that's a thing. At the time I noticed I had about 30 concurrent users queuing/cancelled. I am only a dedicated upload in this case, no searches or downloads. By this point I had 3-4 days in and 1.4TiB upload total. Also other users have said to me that accidentally cancelling their current download from me- cancels the rest of their queued downloads but that may be a separate issue.

**Comment by jpdillingham** [comment]
Thanks for the additional context.

To shed some light on how this is working, search requests come in from the distributed parent at a high rate (like 10/second, guessing) and slskd has to turn around and send them to each child connection immediately.  This works fine if the connection is healthy, but when it isn't stuff starts to back up.  

The last time I went through this I figured out that the pressure was from internal `Task` objects queueing up to send these messages until the connection failed, and then all of the queued Tasks would error in sequence, leading to this sort of memory see-saw pattern.  I added some buffering that would prevent Tasks from queueing too deep, which appeared to solve the problem, but I was aware that the underlying socket and TCP stack on the host machine was still buffering data to send locally before application code was aware there was a problem.  I still don't have a great solution for that because there's a lot of abstraction over the TCP stack and it was made worse when .NET went cross platform.

**Comment by jpdillingham** [comment]
![image](https://user-images.githubusercontent.com/17145758/181914367-22978fa4-9b29-4477-aaf2-bff072e3276a.png)

This is with distributed children disabled.  A couple dozen files were uploaded during this time, and search responses were being sent throughout.  I _think_ this is consistent with the behavior that originally caused me to start down this path, so I think search responses are the culprit rather than distributed children.  It might be both.

I've been doing some profiling locally too:

![image](https://user-images.githubusercontent.com/17145758/181914574-dae77b55-c5fd-4392-ae60-0e33fd0019d6.png)

The biggest contributors to the increase are all GC and framework types:

![image](https://user-images.githubusercontent.com/17145758/181914628-66e046e5-7572-42f3-a06f-2d0ed464c3a6.png)

What I think I'm seeing (locally, anyway) is just the application being greedy with memory (as in, it could deallocate this stuff but chooses not to).  I'm going to play around with a few settings to see if I can stop this behavior, then I'll reevaluate.

**Comment by jpdillingham** [comment]
I set `DOTNET_GCConserveMemory` in the container and it looked promising for the first few hours, but unfortunately allocated memory continues to climb at the rate of about 50mb/day.

I'm still struggling to find ways to get a memory dump from the container without also inflating the image size.  In the meantime I've added the ability to filter incoming search requests, which I can use to filter everything and effectively stop servicing them.  This will give me insight into whether this part of the code is causing the issue.

**Comment by jpdillingham** [comment]
@ckbaudio curious about your config; do you:

* Have `debug` enabled
* Have `soulseek.diagnostic_level` set to `Debug`
* Use the Loki logging integration

I'm seeing pretty drastic differences decreasing logging activities across all platforms.

**Comment by ckbaudio** [comment]
> @ckbaudio curious about your config; do you:
> 
> * Have `debug` enabled
> * Have `soulseek.diagnostic_level` set to `Debug`
> * Use the Loki logging integration
> 
> I'm seeing pretty drastic differences decreasing logging activities across all platforms.

I don't believe I do, at least unless they have been disabled by default. I've noticed on the recent update that memory peaks at around 5 GB and stays pretty steady, but it seems that CPU usage becomes also quite high.

**Comment by jpdillingham** [comment]
Here's the usage for the past 24 hours.  It's been about 72 hours since I turned off detailed diagnostics and disabled logging to Loki.

![image](https://user-images.githubusercontent.com/17145758/182845151-e2f92fff-409e-4217-85b4-b06ad159acde.png)

Around the same time I changed these settings, I also started copies on Windows and Linux desktops, and I saw the same pattern across all three platforms (this one being a Docker container running on an ESXi VM); with debug diagnostics and Loki enabled, memory grows in a linear fashion, with those things disabled it stays flat, with slight variations based on number of children and accounting for any file transfers in flight.

For the time being the official guidance is to leave the `soulseek.diagnostic_level` setting at `Info`, and to not enable the Loki logging integration.  I'll keep working to see if I can understand where the issue is.

Regarding the `VIRT` value from `top`, I _think_ that might include any shared files, which would have been loaded into memory during the initial share scan in order to retrieve the metadata.  I'll do some testing to prove or disprove that theory.

**Comment by jpdillingham** [comment]
@ckbaudio ~600mb RES is much higher than I feel it should be.  When you get the chance, could you please try setting the following environment variables and report back after a few days?

```
  DOTNET_gcServer=0
  DOTNET_gcConcurrent=1
  DOTNET_GCHeapHardLimit=1F400000
  DOTNET_GCConserveMemory=9 
```

I've set these in the Docker container and they might be part of why things are better, but I can't be sure.  These are documented [here](https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector) if you'd prefer to read up before setting them.

**Comment by jpdillingham** [comment]
Here's the last 7 days, running with debugging turned off:

![image](https://user-images.githubusercontent.com/17145758/184381204-84a71854-b1a9-4c05-9f1b-d51da3a53811.png)

I'm satisfied that the memory leak exists only in things related to debugging, and as such the priority of this is greatly reduced.  I still plan to investigate eventually.

**Comment by ckbaudio** [comment]
Hi there,

I haven't enabled any logging for this one. Just occasionally checking htop. I've since updated and had it running for a week and the memory seems to be settling around 750MB, thought CPU load seems high still. This is after 3 TiB shared. 

Will keep on top of it and see if I notice anything odd.

**Comment by jpdillingham** [comment]
This seems to have gone away; the container I use for testing and regression monitoring has had the debug flag on for months and memory hovers around 20mb used.

I noticed when doing some profiling that Serilog uses a ton of memory, accounting for most of the growth in snapshot diffs.  I might see about a replacement that's lighter weight.
